dlib/simd/simd4f_vec.h

```diff
@@ -0,0 +1,104 @@
+// Copyright (C) 2013  Davis E. King (davis@dlib.net)
+// License: Boost Software License   See LICENSE.txt for the full license.
+#ifndef DLIB_SIMD4F_VEC_Hh_
+#define DLIB_SIMD4F_VEC_Hh_
+
+#include "../uintn.h"
+#include "simd4i_vec.h"
+
+namespace dlib
+{
+    typedef simd4i simd4f_bool;
+
+    class simd4f
+    {    
+        typedef union {
+            vector float v;
+            float x[4];
+        } v4f;
+        
+        v4f x;
+        
+    public:
+        inline simd4f() : x{0,0,0,0} {}
+        inline simd4f(const simd4f& v) : x(v.x) { }
+        inline simd4f(const vector float& v) : x{v} { }
+
+        inline simd4f(const simd4i& v) {
+            x.x[0]=v[0]; x.x[1]=v[1]; x.x[2]=v[2]; x.x[3]=v[3];
+        }
+        
+        
+        inline simd4f(float f) : x{f,f,f,f} { }
+        inline simd4f(float r0, float r1, float r2, float r3)
+             : x{r0,r1,r2,r3} { }
+
+        inline simd4f& operator=(const simd4f& v) { x = v.x; return *this; }
+        inline simd4f& operator=(const float& v) { *this = simd4f(v); return *this; }
+
+        inline vector float operator() () const { return x.v; }
```

Comment:
Ok,  they chose to use macros and the compiler doesn't seem to look past the initial type.
An explicit cast works when there is an overloaded cast operator of that type, but it's just as ugly.
I could make a macro LHS(lhs) ((vector T)lhs) and use  vec_macro(LHS,RHS) but I don't know how much better that is than  vec_macro(lhs(),rhs())  or vec_macro((vector T)lhs, (vector T)rhs).

   If you have any other ideas I'll gladly test a better implementation.